### PR TITLE
Prevent Widows for copy items in Onboarding ProcessingScreen

### PR DESCRIPTION
**Pull request by spen**
#### Changes proposed in this Pull Request

This PR aims to prevent the text widows we see in the final screen of signup, right before we're taken to Calypso proper, by adding `<br>` tags to each copy item:

Screenshot 2019-04-22 at 14 37 35

Screenshot 2019-04-29 at 14 11 07

Note that I'm looking to address this in a more robust way by fixing the `preventWidows` util function - I'm tracking that in #32652
For anybody hitting this PR and finding themselves confused by the conversation below, also note that I had originally planned to do that here but then changed the approach to follow @ramonjd's suggestion.

#### Testing instructions

First reproduce to confirm the issue:
- Set your browser viewport to be ~384 pixels wide
- On master & while logged out, go to http://calypso.localhost:3000/start
- Go through the signup steps and watch for the final steps where a large WP logo shows with copy underneath. You should see something like the 'Before' screenshot shown above.

Next, Apply this branch and go through the same process of creating a new account
- This time around, when you hit the last screen, you should see 2 floating words as opposed to one. Just like in the 'After' screenshot above.

### CODE REVIEW

**Comment by spen** [comment]
@dmsnell - I'd love for you to just check over the logic I added if you get time since it's an area you've worked on, but don't stress if you're busy!

**Comment by ramonjd** [comment]
Could we maybe simplify all this by using human power? 

E.g.,

```
				 this.props.translate(
						"{{strong}}Done!{{/strong}} Thanks for waiting, %(domain)s is all set up and we're ready {{br/}}for you to get started.",
						{
							components: { strong: <strong />, br: <br /> },
							args: { domain },
							comment: 'The second line after the breaking tag {{br/}} should fit unbroken in 384px and greater and have a max of 30 characters.'
						}
				  )
```

**Comment by dmsnell** [comment]
Well it didn't turn out simpler and it may not be totally right and I'm definitely skipping the unit tests I asked for but I did some prototyping and think we can consider a different approach based on using `<span style="white-space: nowrap;">…</span>` that treats text and React components the same

https://codesandbox.io/embed/koxqw8x0wo

If you scroll to the bottom of that sandbox you'll find the `<NoWidows minWords={2}>` component and you can adjust the `minWords` parameter and watch it update.

The `<NoWidows />` component could be called `<PreventWidows />` or something else. It gets around the need to surround the `children` in a fragment. If we do what I've done, which is acknowledge that we're working with a mixture of components and text, then we're left in a situation where we need to return an array.

The `0xA0` trick doesn't work with HTML or React components because we can break at the tag boundaries… `&nbsp;<strong>a</strong>&nbsp;` becomes `[ '&nbsp;', <break point>, '<strong>a</strong>', <break point>, '&nbsp;' ]` Therefore I've used the `<span>` to tell the browser not to break instead of modifying the text.

This approach has the benefit that it can preserve the original whitespace, which may have been something like a newline or tab or quad space instead of a normal space, and that can capture certain React components in the wrapping. If the component has few-enough words it will move it into the `nowrap` region and if it's a special few elements, such as a `<strong>` element, then it will split the range into two regions and push one of them into `nowrap`

---

Hope this is helpful. I'd love to see something like this cleaned up and turned into a component with tests and a P2 post about it if you find it useful. Otherwise it's worth adding tests here and some explanatory comments detailing the limitations of this quick adjustment.

**Comment by southp** [comment]
The test scenario works for me. I personally don't have strong preference for a wrapping component or a function, but I'm agreeing with @dmsnell that we should start adding unit tests no matter which direction we go, because it's a complicated problem that is impossible to maintain without good test coverage. Just like the issue of the mixture of pure strings and react components we are addressing here, an even more complicated problem would be i18n -- how could we make our solution general enough for non-alphabetical writings? For example, preventing widows is also a big deal in Chinese, but `/S+/` simply won't work for Chinese, both because it won't match and Chinese doesn't use whitespace characters as separators. 

I know the i18n thing is a much bigger issue that is out of scope, but just wanted to strengthen the point that how complicated this is and how much we need a good test suite :)

**Comment by spen** [comment]
Thanks for looking in to this @dmsnell and @southp :)

In regards to your sandboxed prototype - good work with that! It's given me a lot of hope that we can do a good job at covering strings that include react components beyond what I had covered (only cases where we end with a string).

I did have a play around with your example and I think I spotted some issues which I've tried to point out here: https://codesandbox.io/s/93k74n021y

Note, I'm using `{ false && ( JSX ) }` to sort of comment out code... I find this is a nice quick way to switch between rendering/not rendering JSX.

So you'll see in that example that I've wrapped different parts of the content, rather than all of it. Wrapping all of it would of course only prevent the widows of the last paragraph so I'm wrapping each piece that I would imagine somebody trying in calypso.

When we have nested components - the h2 tag is within the wrapper and contains a strong tag nested within it - we see that we don't prevent the widow as we hoped to.

I've started working on something that attempts to emulate this problem of having nested chunks but, for the sake of simplicity, I'm just looking at nested arrays for the time being - check that out here: https://runkit.com/spen/5cc1815e378595001a1aafaa

I've had some success there so I'm hoping I can apply some of that logic to your methods so we can cover those nested components.

**Comment by dmsnell** [comment]
Thanks for engaging the conversation @southp and @spen.

Let me rephrase my comment a bit to focus more on actionable steps right now, as I think long-term we know this component should be much more robust.

our existing code is undeniably simplistic. it assumes that text is segmented via space characters and that there is only text - totally ineffective with React components

at this stage the problems are self-evident and so they should stick out to any dev who comes around to work in it.

_if_ we start improving the function though by patching over specific use-cases to handle React components then we start to hide the fact that the function is broken. another dev might see the code to handle certain cases and assume that it's more robust/reliable than it really is. worse still without the expected behaviors enumerated anywhere it won't be clear to a passer-by how robust the function is or what it's _supposed_ to do.

_therefore_, my suggestion as we move forward is to start by explicitly showing the range of functionality here via unit tests. they should serve as basic documentation and _in the recognition of the functions limits_ can point that out.

```js
/** @todo this documents the limits of the function but fix it when we make the function better */
it( 'fails to handle general React components until someone fixes it', () => {
	expect( preventWidows( translate( … ) ) ).toBeTotallyWrong();
} );
```

this gives us a little more grant to fix a known case and earmark the function for further refinement later. we don't really have a practice of letting tests fail in `master` so we have to make some compromise if we want to document behaviors in tests.

this could happen differently: in the first PR we add a couple clarifying unit tests showing the text-only use-cases. in a second PR we add a few tests for our added cases

```js
it( 'handles this one case where we have `translate()` sending "text-only" React component output', () => {
	expect( preventWidows( translate( … ) ).toBeRight();
} );
```

that can leave a better audit trail for our thinking when we expanded it.

---

it could make a lot of sense to rewrite the function _now_ to handle basic React usage as in the code sandboxes. limiting the scope of the rewrite is a good move to prevent ourselves from getting stuck for now. I was surprised when I did a cursory search on the internet to find a `<NoWidows>` library and found nothing.

your updates are fine @spen in the prototypes. my main concerns are simply clarifying what we're doing to the next dev who comes around and also not starting with a failed model

**Comment by spen** [comment]
👋 Hey @dmsnell, @southp & @ramonjd - I've gone the route of @ramonjd's suggestion and I'm now tracking the bigger picture solution over in #32652. That'll let us get this one case sorted quickly and hopefully get a more cover-all fix sorted in the future.

I've dismissed your review on this one @dmsnell, now that the approach and code is different, but please feel free to re-review if you have time :)

**Comment by spen** [comment]
Awesome! Appreciate the quick second review @dmsnell 🙏
